File: tools/bench_ports/bench.go
// Run from go/: go run ../tools/bench_ports/bench.go BUNDLE VOICE TEXT
package main

import (
	"encoding/json"
	"fmt"
	"github.com/loudreader/loudkit/go/config"
	"github.com/loudreader/loudkit/go/engine"
	"github.com/loudreader/loudkit/go/onnx"
	"github.com/loudreader/loudkit/go/voice"
	"os"
	"path/filepath"
	"time"
)

func check(err error) {
	if err != nil {
		panic(err)
	}
}
func main() {
	if len(os.Args) != 4 {
		fmt.Fprintln(os.Stderr, "usage: bench BUNDLE VOICE TEXT")
		os.Exit(2)
	}
	bundle := os.Args[1]
	onnx.SetSharedLibraryPath(os.Getenv("LOUDKIT_ONNXRUNTIME_LIB"))
	check(onnx.InitializeEnvironment())
	defer onnx.DestroyEnvironment()
	checkpoint := filepath.Join(bundle, filepath.Base(bundle)+".safetensors")
	start := time.Now()
	eng, err := engine.LoadWith(checkpoint, filepath.Join(bundle, "onnx"), filepath.Join(bundle, "tokenizer.json"), config.ExecutionConfig{ONNXProvider: config.ProviderCPU})
	check(err)
	defer eng.Close()
	load := time.Since(start).Seconds()
	// The checkpoint's rate, not a literal 24000: at any other rate a hardcoded
	// divisor reports the wrong audio duration, and so the wrong RTF, for every run.
	sampleRate := float64(eng.Config().SampleRate)
	v, err := voice.Load(os.Args[2])
	check(err)
	rows := []map[string]any{}
	for run := 0; run < 4; run++ {
		start = time.Now()
		first := 0.
		samples, tokens, chunks := 0, 0, 0
		check(eng.Stream(os.Args[3], v, engine.Options{Seed: 7}, func(c engine.Chunk) bool {
			if chunks == 0 {
				first = time.Since(start).Seconds()
			}
			chunks++
			samples += len(c.Audio)
			tokens += len(c.Tokens)
			return true
		}))
		rows = append(rows, map[string]any{"run": run, "seconds": time.Since(start).Seconds(), "ttfa_s": first, "audio_s": float64(samples) / sampleRate, "tokens": tokens, "chunks": chunks})
	}
	check(json.NewEncoder(os.Stdout).Encode(map[string]any{"runtime": "go", "bundle": bundle, "execution": eng.Describe(), "load_s": load, "text": os.Args[3], "seed": 7, "runs": rows}))
}

File: tools/bench_ports/bench.mjs
// node tools/bench_ports/bench.mjs BUNDLE VOICE TEXT (after npm --prefix js run build)
import { Engine, loadVoice } from '../../js/dist/index.js';
import { performance } from 'node:perf_hooks';
const [bundle, voicePath, text] = process.argv.slice(2);
if (!text) {
  console.error('usage: bench BUNDLE VOICE TEXT');
  process.exit(2);
}
let start = performance.now();
const engine = await Engine.load(bundle, {onnxProvider: 'cpu'});
const load = (performance.now() - start) / 1000;
// The checkpoint's rate, not a literal 24000: at any other rate a hardcoded
// divisor reports the wrong audio duration, and so the wrong RTF, for every run.
const sampleRate = engine.config.sampleRate;
try {
  const voice = loadVoice(voicePath);
  const runs = [];
  for (let run = 0; run < 4; run++) {
    start = performance.now();
    let first = 0, samples = 0, tokens = 0, chunks = 0;
    for await (const c of engine.stream(text, voice, {seed: 7})) {
      if (!chunks) first = (performance.now() - start) / 1000;
      chunks++;
      samples += c.audio.length;
      tokens += c.tokens.length;
    }
    runs.push({
      run,
      seconds: (performance.now() - start) / 1000,
      ttfa_s: first,
      audio_s: samples / sampleRate,
      tokens,
      chunks,
    });
  }
  console.log(JSON.stringify({runtime: 'js', bundle, load_s: load, text, seed: 7, runs}));
} finally {
  await engine.close();
}
